src/main/java/jenkins/plugins/openstack/compute/JCloudsSlaveTemplate.java

```diff
@@ -227,10 +230,59 @@ public boolean canProvision(final Label label) {
         // Ensure predictable server name so we can inject it into user data
         builder.name(serverName);
 
-        if (!Strings.isNullOrEmpty(opts.getImageId())) {
-            String imageId = cloud.getOpenstack().getImageIdFor(opts.getImageId());
-            LOGGER.fine("Setting image id to " + imageId);
-            builder.image(imageId);
+        final Openstack openstack = cloud.getOpenstack();
+        final JCloudsCloud.BootSource bootSource = opts.getBootSource() == null ? JCloudsCloud.BootSource.IMAGE : opts.getBootSource();
+        final String imageNameOrId = opts.getImageId();
+        final String effectiveImageId;
+        if (!Strings.isNullOrEmpty(imageNameOrId)) {
+            final List<String> ids;
+            switch (bootSource) {
+                case IMAGE :
+                    ids = openstack.getImageIdsFor(imageNameOrId);
+                    break;
+                case VOLUMESNAPSHOT :
+                    ids = openstack.getVolumeSnapshotIdsFor(imageNameOrId);
+                    break;
+                default :
+                    LOGGER.config("Don't know how to handle bootSource of " + bootSource + ".");
+                    ids = Collections.EMPTY_LIST;
+                    break;
+            }
+            final int numberFound = ids.size();
+            switch (numberFound) {
+                case 0 :
+                    LOGGER.warning("No active " + bootSource + " '" + imageNameOrId + "' could be found.");
+                    effectiveImageId = null;
+                    break;
+                default :
+                    effectiveImageId = ids.get(numberFound - 1);
+                    LOGGER.warning("Multiple " + bootSource + "s (" + numberFound + ") found with name '" + imageNameOrId
+                            + "'.  Using most recent one, '" + effectiveImageId + "'.");
```

Comment:
Is there any other reason why this API returns a collection of images/snapshots except for this message logged (`getImageIdFor` return id or null)? If there is not, I am ok to revert the semantics back and do logging inside. It seems to simplify both client and implementation.